### PR TITLE
executor: fix hash join sqllogic test block

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes CI failure introduced by  #8338

### What is changed and how it works?

- don't use waitGroup
- listen closeChan close
- listen innerResultChan close

### Check List 

Tests 

 - Unit test

Code changes

 - impl change

Side effects

 - N/A

Related changes

 - N/A

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8374)

### CODE REVIEW

**Comment by lysu** [comment]
@zz-jason  @tiancaiamao

**Comment by zz-jason** [comment]
LGTM

**Comment by tiancaiamao** [comment]
Please add test for the bug fix.
You may construct the blocked SQL, or mock using gofail.
@lysu

File: executor/join.go [hunk 1]
```diff
@@ -262,10 +261,7 @@ func (e *HashJoinExec) wait4Inner() (finished bool, err error) {
 // fetchInnerRows fetches all rows from inner executor,
 // and append them to e.innerResult.
 func (e *HashJoinExec) fetchInnerRows(ctx context.Context, chkCh chan<- *chunk.Chunk, doneCh chan struct{}) {
-	defer func() {
-		close(chkCh)
-		e.fetchInnerDone.Done()
-	}()
+	defer close(chkCh)
```
Comment (eurekaka) [inline executor/join.go]: Could you explain why `WaitGroup` does not work? thanks

Comment (lysu) [reply executor/join.go]: this exec is too complex, the situation missed by us is that `finished` flag will make `chkCh`'s consumer goroutine exit before producer one, lastday we found producer are blocked on `chkCh <- chk` then no chance to call `wg.Done` 

maybe it's more suitable consume data(also check closeCh) and channel close to handle this...

**Comment by tiancaiamao** [comment]
LGTM
We have to fix CI first, `join.go` need to be carefully reviewed later.

**Pull request merged**